# BME205 - Bioinformatics Models and Algorithms
This course was my first graduate course at UCSC, and was a huge learning experience. Being an undergrad in a graduate course can be rather daunting, but the material made me feel at home. This course combines concepts from three courses in UCSC, CMPS 101 (mabdulqa/CMPS101) - an ADT and Algorithms course that introduced students graph theory and the algorithm heavy portions of computer science, BME 160 (mabdulqa/BME160) - an introductory Python programming course for applications in the fields of molecular biology/biochemistry/evolutionary biology/bioinformatics, and AMS131 (no repository here - it is a statistics course) - an introduction to probability theory and frequentist statistics and most importantly markov chains. Using elements from the 3 courses - BME205 introduced the foundational alogrihtms and statistical/mathematical models that are essentailly the back bone of bioinformatics. The course began with finding hidden motifs using a bayesian approach, then to popular geneome assembly algorithms like DeBruijn and Eulerian Path, to sequence alignment, to applications of hidden markov models - mainly the forward backward alogrithm and Baum-Welch learning.
 
